syntax = "proto3"; // the version at the time of writing

package grpc;

message Empty {}

message Message {
    string username = 1;
    string recipient = 2;
    string message = 3;
}

message SignupRequest {
    string username = 1;
}

message SignupReply {
    bool success = 1;
    string error = 2;
}

message LoginRequest {
    string username = 1;
}

message LoginReply {
    bool success = 1;
    string error = 2;
}

message LogoutRequest {
    string username = 1;
}

message LogoutReply {
    bool success = 1;
    string error = 2;
}

service ChatServer {
    // This bi-directional stream makes it possible to send and receive Notes between 2 persons
    rpc ChatStream (Empty) returns (stream Message); 
    rpc SendMessage (Message) returns (Empty);
    rpc Signup (SignupRequest) returns (SignupReply);
    rpc Login (LoginRequest) returns (LoginReply);
    rpc Logout (LogoutRequest) returns (LogoutReply);
}